tunnel/intra/doh/client_auth_test.go

```diff
@@ -0,0 +1,378 @@
++qA3p6a0zqx89dT4jv14qsKSWPMju9kGqaS/+0WFfrsPYSX1PmDhN7WU
+-----END PRIVATE KEY-----`
+
+// PEM encoded test leaf certificate with RSA public key.
+// Doubles as an intermediate depending on the test.
+var rsaCertificate string = `-----BEGIN CERTIFICATE-----
+MIICWDCCAcGgAwIBAgIUS36guwZMKNO0ADReGLi0cZq8fOowDQYJKoZIhvcNAQEL
+BQAwPjELMAkGA1UEBhMCVVMxCzAJBgNVBAgMAkNBMRYwFAYDVQQHDA1Nb3VudGFp
+biBWaWV3MQowCAYDVQQKDAFYMB4XDTIwMTEwNDA1NDgyNVoXDTMwMTEwMjA1NDgy
+NVowPjELMAkGA1UEBhMCVVMxCzAJBgNVBAgMAkNBMRYwFAYDVQQHDA1Nb3VudGFp
+biBWaWV3MQowCAYDVQQKDAFYMIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQDd
+eznqVu1Rn0m8KR4mX/qVv6uytzZ+juqW5VD55D+w9N6JryPpFHPi4VIm8PKLXp3X
+GvY9mc8r+0Ow1qJZYoc/X0Na1c79bv9xwbD3aK28FlAs1+cmyesaFhCWa0bYAvcy
+mqQGYhObEWb46E5AANV82CitDE9C1aXRT4SvkLnc6wIDAQABo1MwUTAdBgNVHQ4E
+FgQUnUib8BhOHqjq9+gqPQ+ePyEW9zwwHwYDVR0jBBgwFoAUnUib8BhOHqjq9+gq
+PQ+ePyEW9zwwDwYDVR0TAQH/BAUwAwEB/zANBgkqhkiG9w0BAQsFAAOBgQAx/uZG
+Gmb5w/u4UkdH7wnoOUNx6GwdraqtQWnFaXb87PmuVAjBwSAnzes2mlp/Vbcd6tYs
+pPuHrxOcWgw/aRV6rK3vJZIH3DGvy1pNphGgegEcG88nrUCDcQqPLxvPJ8bmbaee
+Tf+l5U2OHC3Yifb4FDOv47kGmq5VeWiYdp60/A==
+-----END CERTIFICATE-----`
+
+// PKCS8 encoded test RSA private key.
+var rsaKey string = `-----BEGIN PRIVATE KEY-----
+MIICeAIBADANBgkqhkiG9w0BAQEFAASCAmIwggJeAgEAAoGBAN17OepW7VGfSbwp
+HiZf+pW/q7K3Nn6O6pblUPnkP7D03omvI+kUc+LhUibw8otendca9j2Zzyv7Q7DW
+ollihz9fQ1rVzv1u/3HBsPdorbwWUCzX5ybJ6xoWEJZrRtgC9zKapAZiE5sRZvjo
+TkAA1XzYKK0MT0LVpdFPhK+QudzrAgMBAAECgYEAoCdhI8Ej7qe+S993u8wfiXWG
+FL9DGpUBsYe03F5eZ/lJikopL3voqKDCJQKKgJk0jb0jXjwAgQ86TX+G+hezL5jp
+xOOfMmTYgMwnUuFYN1gHAd+TnYB9G1qSQr9TOw3K9Rf4q2x09GhLP75qdr+qzmIR
+YGle5ZSP0LqKNkpGNUECQQD+6CxOO8+knnzIFvqkUyNDVFR5ALRNpb53TGVITNf3
+ysT32oJ75ButA0l4q/jsL+MeLLvrHkJOHN+ydLaZOUkbAkEA3m5cICisW9lsT+Rj
+glXykkbj3Ougldy7rhPivAaS7clk8cl8cDcIvHna1mDlhSanUu/s4TFEXBLnSzee
+XLNIcQJBAJ0n3TD6lSEkCUB/UlX/X81B77aOZZs9pXj9o6/4mGoQHHHGyQ3C7AE1
+9pUsSZKsT3UqFU124WAxUwU+CdnbxKMCQB/QrUC0UKL6oHF0+37DCGU/2ovY8Ck/
+X2Dw2zeFwTJd4iBrb28lkAxVaaXMSkgXVUuZoco8H8kDsy2hEPe1dSECQQCPw5Yg
+2gdmdpUk+QetqqhSuuIDwILHU9m3CoX3rY+njaR5LOWDz3utC9Ogo+4wdIMamP/o
+2SAWPAZPqDUbtqGH
+-----END PRIVATE KEY-----`
+
+// fakeClientAuth implements the ClientAuth interface for testing.
+type fakeClientAuth struct {
+	ClientAuth
+
+	certificate  *x509.Certificate
+	intermediate *x509.Certificate
+	key          crypto.PrivateKey
+}
+
+func (ca *fakeClientAuth) GetClientCertificate() []byte {
+	if ca.certificate == nil {
+		// Interface uses nil for errors to support binding.
+		return nil
+	}
+	return ca.certificate.Raw
+}
+
+func (ca *fakeClientAuth) GetIntermediateCertificate() []byte {
+	if ca.intermediate == nil {
+		return nil
+	}
+	return ca.intermediate.Raw
+}
+
+func (ca *fakeClientAuth) Sign(digest []byte) []byte {
+	if ca.key == nil {
+		return nil
+	}
+	if k, isECDSA := ca.key.(*ecdsa.PrivateKey); isECDSA {
+		// We have to ASN.1 encode the signature ourselves since
+		// Go < 1.15 does not include ecdsa.SignASN1
```

Comment:
outline-go-tun2socks now requires go 1.15 in go.mod, so we should be able to use that.

Reply:
Excellent.